### PR TITLE
H3Hexagon: add support for coverage.

### PR DESCRIPTION
**Pull request by 1chandu**
For #3209 

#### Background
- Forward `coverage` prop to `ColumnLayer`
- Scale polygon vertices when using `PolygonLayer`.

#### Change List
- H3Hexagon: add support for coverage.

H3Coverage

### CODE REVIEW

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 1]
```diff
@@ -22,9 +23,27 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
   return [lng, lat];
 }
 
-function h3ToPolygon(hexId) {
+function scalePolygon(hexId, vertices, factor) {
+  const [lat, lng] = h3ToGeo(hexId);
+  // get distinct vertex objects,
+  // `h3ToGeoBoundary` returns same array object for first and last vertex (closed polygon)
+  const uniqueVertices = [...new Set(vertices)];
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: - Don't use spread operator in this repo
- This is unnecessary. the construction of new objects is more expensive than transforming one additional point. `h3ToPolygon` should consistently return the same format regardless of `coverage`.

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 2]
```diff
@@ -22,9 +23,27 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
...
+  for (const vertex of uniqueVertices) {
+    const scaledPt = lerp([lng, lat], vertex, factor);
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: Do not create new array in for loop. You can simply call `lerp` twice.

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 3]
```diff
@@ -36,6 +55,18 @@ function h3ToPolygon(hexId) {
   return vertices;
 }
 
+function mergeTriggers(currentTrigger, newTrigger) {
+  let trigger;
+  if (!currentTrigger) {
+    trigger = newTrigger;
+  } else if (Array.isArray(currentTrigger)) {
+    trigger = [...currentTrigger, newTrigger];
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: Perf: use `concat`

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 4]
```diff
@@ -36,6 +55,18 @@ function h3ToPolygon(hexId) {
...
+  } else {
+    trigger = [currentTrigger, newTrigger];
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: `currentTrigger` can be an object.

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js (hunk 1 above)
Comment (1chandu) [reply modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: `h3ToGeoBoundary` is returning same object (Array object) for first and last vertex , changing one changes the other, I will have WA here (just skip the last), and will open bug for h3-js.

Comment (Pessimistress) [reply modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: I see. I don't think this is necessarily a bug for h3-js, but indeed a catch.

**Comment by 1chandu** [comment]
@Pessimistress , addressed comments.

deck.gl 7.0 supported the coverage for non `highPrecision` mode (we were forwarding the prop), any concerns cherry-picking this into 7.1-release branch ?

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 5]
```diff
@@ -22,9 +23,26 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
   return [lng, lat];
 }
 
-function h3ToPolygon(hexId) {
+function scalePolygon(hexId, vertices, factor) {
+  const [lat, lng] = h3ToGeo(hexId);
+  // `h3ToGeoBoundary` returns same array object for first and last vertex (closed polygon),
+  // hence skip scaling the last vertex
+  log.assert(vertices[0] === vertices[vertices.length - 1]);
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: Perf: save `vertices.length` in a variable:

```
const n = vertices.length;
for (let i = 1; i < n; i++) {
  ...
}
```

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 6]
```diff
@@ -36,6 +54,20 @@ function h3ToPolygon(hexId) {
   return vertices;
 }
 
+function mergeTriggers(currentTrigger, newTrigger) {
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: What's the point of creating a `newTrigger` object if you're just going to use `newTrigger.cover`?

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 7]
```diff
@@ -36,6 +54,20 @@ function h3ToPolygon(hexId) {
...
+  let trigger;
+  if (currentTrigger === undefined || currentTrigger === null) {
+    trigger = newTrigger.coverage;
+  } else if (Array.isArray(currentTrigger)) {
+    trigger = [].concat(currentTrigger, newTrigger.coverage);
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: `Object.assign({}, <array>)` works, you don't have to handle it as a special case.

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 8]
```diff
@@ -36,6 +54,20 @@ function h3ToPolygon(hexId) {
...
+  } else if (typeof currentTrigger === 'object') {
+    trigger = Object.assign({}, currentTrigger, newTrigger);
+  } else {
+    trigger = [currentTrigger, newTrigger.coverage];
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: `{currentTrigger, coverage}`

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 9]
```diff
@@ -22,9 +23,26 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
   return [lng, lat];
 }
 
-function h3ToPolygon(hexId) {
+function scalePolygon(hexId, vertices, factor) {
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: Add a test for this function?

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js (hunk 6 above)
Comment (1chandu) [reply modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: Either I have to pass value and do {coverage: value} or pass an object and access it as object.coverage. Changed it to latter former case.

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 10]
```diff
@@ -22,9 +23,27 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
...
+  }
+}
+
+function h3ToPolygon(hexId, coverage = 1) {
   const vertices = h3ToGeoBoundary(hexId, true);
   const refLng = vertices[0][0];
+
+  if (coverage !== 1) {
+    // updates array object elements of vertices array.
+    scalePolygon(hexId, vertices, coverage);
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: This breaks if the hexagon crosses the 180th meridian. You need to set the refLng to that of the center, normalize it before scaling.

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 11]
```diff
@@ -22,9 +23,27 @@ function getHexagonCentroid(getHexagon, object, objectInfo) {
   return [lng, lat];
 }
 
-function h3ToPolygon(hexId) {
+export function scalePolygon(hexId, vertices, factor) {
+  const [lat, lng] = h3ToGeo(hexId);
+  const vertexCount = vertices.length;
+  log.assert(vertices[0] === vertices[vertexCount - 1]);
```
Comment (Pessimistress) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: This assert will break the layer of h3-js makes a change in the future. It won't be the user's fault and it won't provide any useful error message. Can we just adapt the for loop based on whether these two points are the same object?

File: test/modules/geo-layers/h3-layers.spec.js [hunk 12]
```diff
@@ -50,6 +54,121 @@ test('H3HexagonLayer', t => {
   t.end();
 });
 
+test.only('H3HexagonLayer#mergeTriggers', t => {
```
Comment (Pessimistress) [inline test/modules/geo-layers/h3-layers.spec.js]: test.only

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js (hunk 10 above)
Comment (1chandu) [reply modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: Done.

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js (hunk 11 above)
Comment (1chandu) [reply modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: Done.

File: modules/geo-layers/src/h3-layers/h3-hexagon-layer.js [hunk 13]
```diff
@@ -33,9 +28,57 @@ function h3ToPolygon(hexId) {
...
+function h3ToPolygon(hexId, coverage = 1) {
+  const vertices = h3ToGeoBoundary(hexId, true);
+
+  if (coverage !== 1) {
+    // updates array object elements of vertices array.
+    scalePolygon(hexId, vertices, coverage);
+  }
+
+  // normalize with respect start vertex
+  normalizeLongitudes(vertices);
```
Comment (1chandu) [inline modules/geo-layers/src/h3-layers/h3-hexagon-layer.js]: I wonder we should move this into `PolygonLayer` or it's sublayers.

**Pull request merged**